### PR TITLE
[RFC] Upgrade Typescript

### PR DESCRIPTION
**Pull request by johnjeng**
# Description
Upgrade TS to 3.1.6

# Testing Steps
1. Added students to courses
1. Distributed assignments to students
1. Create and edited txt, ts
1. Ran `2+2` in a Jupyter notebook

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):
- [x] No debugging console.log messages.
- [x] All new code is actually used.
- [x] Non-obvious code has some sort of comments.

Front end:
- [x] Restart at least one project and check its `~/.smc/local_hub/local_hub.log`
- [x] Completely restart Webpack with `./w` in `/src`
- [x] Screenshots if relevant.

### CODE REVIEW

File: src/webapp-lib/app.pug [hunk 1]
```diff
@@ -26,8 +26,8 @@ html(lang="en")
       | Initializing ...
 
     //- preflight check comes **after** the startup banner and status info, and before any code that might be syntactically broken.
-    script(type="text/javascript")
-      include:typescript preflight-checks.ts
+    //script(type="text/javascript")
+    //  include:typescript preflight-checks.ts
```
Comment (johnjeng) [inline src/webapp-lib/app.pug]: @haraldschilly Whenever you have time, can you check why this fails? Just uncomment.

Comment (haraldschilly) [reply src/webapp-lib/app.pug]: I checked this branch out, updated, made sure I work with the updated version, uncommented it:

```
~$ which tsc
/home/user/smc/src/node_modules/.bin/tsc
~$ tsc --version
Version 3.1.6
```

but the app.html page builds fine and the preflight check comment in the js log shows up with the correct version number. what exactly is going wrong? maybe I have something cached somewhere, though.

**Comment by johnjeng** [comment]
This appears to work for me as well now. @DrXyzzy can you run some Jupyter tests?

**Comment by haraldschilly** [comment]
This pull request **introduces 2 alerts** and **fixes 2** when merging 1bd3e3243bc40975e42949807c32935cff7d633f into 0d35f5f67de5bb9cad5e64c9db9011a4bb949c73 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-d413b93d6658315a259bf5e4d280f5f515b2a8aa)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Self assignment

**fixed alerts:**

* 2 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by williamstein** [comment]
This is perhaps going to be a difficult PR.  

1. [ ] **The hub images:** Building works, but running fails in Kubernetes.  Doing `co hub init -t now` works, but the hubs do not start running.  They all crash on startup with this branch, as follows:
```
Error: Cannot find module 'typescript'
    at Function.Module._resolveFilename (module.js:548:15)    at Function.Module._load (module.js:475:25)    at Module.require (module.js:597:17)    at require (internal/module.js:11:18)
    at Object.register (/cocalc/src/node_modules/ts-node/src/index.ts:227:26)
    at Object.<anonymous> (/cocalc/src/smc-hub/hub.coffee:40:22)
    at Object.<anonymous> (/cocalc/src/smc-hub/hub.coffee:963:4)
    at Module._compile (module.js:653:30)
    at Object.require.extensions..coffee (/cocalc/src/smc-hub/node_modules/coffee2-cache/index.js:84:19)
    at Module.load (/cocalc/src/smc-hub/node_modules/coffeescript/lib/coffeescript/register.js:53:36)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/cocalc/src/smc-hub/run/hub.js:3:1)
    at Module._compile (module.js:653:30)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Function.Module.runMain (module.js:694:10)
    at startup (bootstrap_node.js:204:16)
    at bootstrap_node.js:625:3
```

Your checklist above doesn't include just stopping the hub and starting it in a project.   Can you add that (in addition to restarting a project, etc.)?  

2. [ ] **Building the static image** (webpack).  This fails with:
```
[1964] ./webapp-css.coffee 1.51 KiB {4} [built]
[1967] multi ./webapp-lib/primus/primus-engine.min.js 28 bytes {1} [built]
    + 1954 hidden modules

ERROR in /cocalc/src/smc-webapp/smc_chat.tsx
./smc-webapp/smc_chat.tsx
[tsl] ERROR in /cocalc/src/smc-webapp/smc_chat.tsx(682,10)
      TS2741: Property '[Invalid_propTypes]' is missing in type '{ key: any; account_id: string; history: any; user_map: any; message: any; date: any; project_id: string | undefined; file_path: string | undefined; font_size: number | undefined; is_prev_sender: boolean; ... 9 more ...; set_scroll: Function | undefined; }' but required in type 'Invalid_propTypes<Readonly<{ children?: ReactNode; }> & Readonly<MessageProps>, { actions: any; focus_end: any; get_user_name: any; message: any; history: any; account_id: any; date: any; sender_name: any; ... 9 more ...; saved_mesg: any; }>'.
Child html-webpack-plugin for "app.html":
     5 assets
```

3. [x]  **Building the project image**: It builds fine. And projects *do work* fine.

4. [ ] Manage pods.  I'll wait until 1 is resolved, then try this again...

**Comment by haraldschilly** [comment]
ad "Building the static image": does this have the recent commits regarding travis-ci? there is one build target that simulates what's going on in kucalc. I'll check and if not, I'll push a merge with master.

**Comment by williamstein** [comment]
>  does this have the recent commits regarding travis-ci?

No -- it was using the kucalc env from 20 days ago (test-op).  Can you try building instead of me, since I'm working on other PR's now?  Thanks!

**Comment by haraldschilly** [comment]
After upstream-merging this branch, I waited 10 minutes and travis caught it: https://travis-ci.org/sagemathinc/cocalc/jobs/461921057#L868 … very good

**Comment by williamstein** [comment]
> very good

Nice!

**Comment by haraldschilly** [comment]
This pull request **introduces 2 alerts** and **fixes 2** when merging 9a4f749a1bee15de0220a7ea7926646ed3f64963 into 819edb097d2a9f8f9ba0f697c42386ed73d42047 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-07516c4302afd2cb1de013654efa8d835206d294)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Self assignment

**fixed alerts:**

* 2 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by johnjeng** [comment]
Hmmm https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30957 could be relevant. Going to try using the specific versions that work.

**Comment by haraldschilly** [comment]
This pull request **introduces 2 alerts** and **fixes 2** when merging 9bda42d88b21e9fc55f93a6cc63cb4fc460eb428 into 3f1449742a6ec701bbccf1fe3a933d9816ee28b1 - [view on LGTM.com](https://lgtm.com/projects/g/sagemathinc/cocalc/rev/pr-dd85949da1288e15b755d2d967b05c49e988b8bc)

**new alerts:**

* 1 for Useless assignment to local variable
* 1 for Self assignment

**fixed alerts:**

* 2 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by johnjeng** [comment]
Uhhh passes Travis ci after formatting with Prettier?

**Comment by williamstein** [comment]
We can use [this](https://github.com/bterlson/strict-event-emitter-types) after we get this PR merged...

**Comment by johnjeng** [comment]
This seemed to work when I ran it on test.cocalc.com last.

**Comment by williamstein** [comment]
We can use [this](https://github.com/bterlson/strict-event-emitter-types) after we get this PR merged...

**Comment by williamstein** [comment]
> This seemed to work when I ran it on test.cocalc.com last.

Which test builds did you run? 

- [ ] `co static init -t now`
- [ ] `co hub init -t now`
- [ ]  `co manage init -t now`

**Comment by johnjeng** [comment]
Ah, I didn't catch the other test builds.

**Pull request merged**